kv/kvnemesis: TestKVNemesisMultiNode failed

kv/kvnemesis.TestKVNemesisMultiNode [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/11969886?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/11969886?buildTab=artifacts#/) on master @ [6e49f7ff2955f01c3710a1dec7ad4d9e06efb00a](https://github.com/cockroachdb/cockroach/commits/6e49f7ff2955f01c3710a1dec7ad4d9e06efb00a):


```
=== RUN   TestKVNemesisMultiNode
    test_log_scope.go:167: test logs captured to: /artifacts/tmp/_tmp/4aeb9c412c16a003ed8ddc43788d2bba/logTestKVNemesisMultiNode2713155144
    test_log_scope.go:81: use -show-logs to present logs inline
    kvnemesis_test.go:283: seed: 670630703469535428
    kvnemesis_test.go:197: kvnemesis logging to /artifacts/tmp/_tmp/4aeb9c412c16a003ed8ddc43788d2bba/kvnemesis3964384411
    kvnemesis.go:168: error applying x.Txn(ctx, func(ctx context.Context, txn *kv.Txn) error {
          txn.SetIsoLevel(isolation.Snapshot)
          txn.ReverseScanForShareSkipLocked(ctx, tk(3546959026506480144), tk(13910507768101481354), 0) // SKIP LOCKED request should not find another waiting request from the same transaction
          txn.GetForShare(ctx, tk(4004235856726145791)) // omitted
          b := &kv.Batch{}
          b.ScanForShareGuaranteedDurability(tk(3017471283909807992), tk(6960863703333348775)) // omitted
          b.ScanForShareGuaranteedDurability(tk(14928907874121331292), tk(16880014683745090268)) // omitted
          txn.CommitInBatch(ctx, b) // omitted
          return nil
        }) // SKIP LOCKED request should not find another waiting request from the same transaction: SKIP LOCKED request should not find another waiting request from the same transaction
    kvnemesis.go:168: error applying x.ReverseScanForShareSkipLocked(ctx, tk(3546959026506480144), tk(13910507768101481354), 0) // SKIP LOCKED request should not find another waiting request from the same transaction: SKIP LOCKED request should not find another waiting request from the same transaction
    kvnemesis.go:203: failures(verbose): /artifacts/tmp/_tmp/4aeb9c412c16a003ed8ddc43788d2bba/kvnemesis3964384411/failures
        repro steps: /artifacts/tmp/_tmp/4aeb9c412c16a003ed8ddc43788d2bba/kvnemesis3964384411/repro.go
        rangefeed KVs: /artifacts/tmp/_tmp/4aeb9c412c16a003ed8ddc43788d2bba/kvnemesis3964384411/kvs-rangefeed.txt
        scan KVs: /artifacts/tmp/_tmp/4aeb9c412c16a003ed8ddc43788d2bba/kvnemesis3964384411/kvs-scan.txt
    kvnemesis_test.go:310: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/kv/kvnemesis/pkg/kv/kvnemesis/kvnemesis_test.go:310
        	            				github.com/cockroachdb/cockroach/pkg/kv/kvnemesis/pkg/kv/kvnemesis/kvnemesis_test.go:261
        	Error:      	Should be zero, but was 2
        	Test:       	TestKVNemesisMultiNode
        	Messages:   	kvnemesis detected failures
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/4aeb9c412c16a003ed8ddc43788d2bba/logTestKVNemesisMultiNode2713155144
--- FAIL: TestKVNemesisMultiNode (13.57s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestKVNemesisMultiNode.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-31893

--- a/pkg/kv/kvnemesis/applier.go
+++ b/pkg/kv/kvnemesis/applier.go
@@ -115,6 +115,10 @@ func exceptSharedLockPromotionError(err error) bool { // true if lock promotion 
 	return errors.Is(err, &concurrency.LockPromotionError{})
 }
 
+func exceptSkipLockedReplayError(err error) bool { // true if skip locked replay error
+	return errors.Is(err, &concurrency.SkipLockedReplayError{})
+}
+
 func applyOp(ctx context.Context, env *Env, db *kv.DB, op *Operation) {
 	switch o := op.GetValue().(type) {
 	case *GetOperation,

--- a/pkg/kv/kvnemesis/validator.go
+++ b/pkg/kv/kvnemesis/validator.go
@@ -705,7 +705,10 @@ func (v *validator) processOp(op Operation) {
 		//
 		// So we ignore the results of failIfError, calling it only for its side
 		// effect of perhaps registering a failure with the validator.
-		v.failIfError(op, t.Result, exceptRollback, exceptAmbiguous, exceptSharedLockPromotionError)
+		v.failIfError(
+			op, t.Result,
+			exceptRollback, exceptAmbiguous, exceptSharedLockPromotionError, exceptSkipLockedReplayError,
+		)
 
 		ops := t.Ops
 		if t.CommitInBatch != nil {
@@ -1208,6 +1211,7 @@ func (v *validator) checkError(
 		exceptAmbiguous, exceptOmitted, exceptRetry,
 		exceptDelRangeUsingTombstoneStraddlesRangeBoundary,
 		exceptSharedLockPromotionError,
+		exceptSkipLockedReplayError,
 	}
 	sl = append(sl, extraExceptions...)
 	return v.failIfError(op, r, sl...)

--- a/pkg/kv/kvserver/concurrency/lock_table.go
+++ b/pkg/kv/kvserver/concurrency/lock_table.go
@@ -747,9 +747,22 @@ func (g *lockTableGuardImpl) IsKeyLockedByConflictingTxn(
 			break
 		}
 		if g.isSameTxn(qqg.guard.txnMeta()) {
-			return false, nil, errors.AssertionFailedf(
-				"SKIP LOCKED request should not find another waiting request from the same transaction",
-			)
+			// A SKIP LOCKED request should not find another waiting request from its
+			// own transaction, at least not in the way that SQL uses KV. The only way
+			// we can end up finding another request in the lock's wait queue from our
+			// own transaction is if we're a replay.  We could handle this case by
+			// treating it as a non-conflict, but doing so expands the testing surface
+			// area -- we would want to include tests for:
+			// 1. Just our own request in the wait queue, treated as a non-conflict.
+			// 2. A request from a different transaction in the wait queue, with a
+			// lower sequence number, that conflicts.
+			// 3. A request from a different transaction in the wait queue, with a
+			// higher sequence number, that conflicts.
+			// 4. A request from a different transaction in the wait queue, with a
+			// lower sequence number, that does not conflict.
+			// For now, we simply return an error, and mark it for the benefit of
+			// KVNemesis.
+			return false, nil, MarkSkipLockedReplayError(errors.Errorf("SKIP LOCKED request should not find another waiting request from the same transaction"))
 		}
 		if lock.Conflicts(qqg.mode, makeLockMode(str, g.txnMeta(), g.ts), &g.lt.settings.SV) {
 			return true, nil, nil // the conflict isn't with a lock holder, nil is returned
@@ -4277,4 +4290,22 @@ func MarkLockPromotionError(cause error) error {
 		return nil
 	}
 	return errors.Mark(cause, &LockPromotionError{})
+}
+
+// SkipLockedReplayError is used to mark errors resulting from replayed SKIP
+// LOCKED requests that discover other requests from their own transactions in
+// a lock's wait queue. We mark such errors for the benefit of KVNemesis.
+type SkipLockedReplayError struct{}
+
+func (e *SkipLockedReplayError) Error() string {
+	return "skip locked replay error"
+}
+
+// MarkSkipLockedReplayError wraps the given error, if non-nil, as a skip locked
+// replay error.
+func MarkSkipLockedReplayError(cause error) error {
+	if cause == nil {
+		return nil
+	}
+	return errors.Mark(cause, &SkipLockedReplayError{})
 }

